assignment-3/src/data/ArrayListStudentDao.java

```diff
@@ -0,0 +1,78 @@
+package data;
+
+import logic.Student;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+
+public class ArrayListStudentDao implements data.StudentDao {
+    private final List<Student> students = new ArrayList<>();
+
+    @Override
+    public void add(Student student) throws Exception {
+        for (Student s : students) {
+            if (s.getId() == student.getId()) {
+                throw new Exception("Student with ID " + student.getId() + " already exists.");
+            }
+        }
+        students.add(student);
+    }
+
+    @Override
+    public void update(Student student) {
+        Student s = getById(student.getId());
+        if (s != null) {
+            s.setFirstName(student.getFirstName());
+            s.setLastName(student.getLastName());
+        }
+        else{
+            throw new NoSuchStudentException(student.getId());
+        }
+    }
+
+    @Override
+    public void delete(int id) {
+        Student s = getById(id);
+        if (s != null) {
+            students.remove(s);
+        } else {
+            throw new NoSuchStudentException(id);
+        }
+    }
+
+    @Override
+    public Student getById(int id) {
```

Comment:
Good implementation, try to use the Java Stream API: https://www.baeldung.com/java-8-streams

Here you can use the methods `filter`, `findFirst`, and `orElseThrow` of `stream`.